WIP: Add docs for Cloudflare adapter (#3)
opennextjs/docs: pages/cloudflare/bindings.mdx

import { SITE } from '../../config';
import { Callout } from 'nextra/components';

### Bindings

[Bindings](https://developers.cloudflare.com/workers/runtime-apis/bindings/) allow your Worker to interact with resources on the Cloudflare Developer Platform. When you declare a binding on your Worker, you grant it a specific capability, such as being able to read and write files to an [R2](https://developers.cloudflare.com/r2/) bucket.

<Callout>
If your app uses both bindings and @opennextjs/cloudflare, you currently cannot access bindings when running `next dev`. Instead, you must build your app using @opennextjs/cloudflare and then run it locally in the Workers runtime as described [here](/cloudflare/get-started). This means you cannot use hot-module reloading (HMR) for a faster development workflow. @opennextjs/cloudflare is pre 1.0, in active development, and we plan to address this soon.

Alternatively, you can use [@cloudflare/next-on-pages](https://www.npmjs.com/package/@cloudflare/next-on-pages) to deploy Next.js apps to Cloudflare Pages. You can review the differences in supported Next.js features between @opennextjs/cloudflare and @cloudflare/next-on-pages [here](LINK).

> </Callout>

Reviewer: This callout should be removed as support for bindings in `next dev` has been added here: https://github.com/opennextjs/opennextjs-cloudflare/pull/33

#### How to configure your Next.js app so it can access bindings

> Install [@opennextjs/cloudflare](https://www.npmjs.com/package/@opennextjs/cloudflare), and then add a `wrangler.toml` file in the root directory of your Next.js app, as described in [Get Started](/cloudflare/get-started), y

Reviewer 2: ```suggestion
Install [@opennextjs/cloudflare](https://www.npmjs.com/package/@opennextjs/cloudflare), and then add a `wrangler.toml` file in the root directory of your Next.js app, as described in [Get Started](/cloudflare/get-started)
```

#### How to access bindings in your Next.js app

> You can access [bindings](https://developers.cloudflare.com/workers/runtime-apis/bindings/) from any route of your Next.js app via `getRequestContext`:

Reviewer: ```suggestion
You can access [bindings](https://developers.cloudflare.com/workers/runtime-apis/bindings/) from any route of your Next.js app via `getCloudlfareContext`:
```

```js
// TODO: possible to bring this in from next-on-pages?

> import { getRequestContext } from "@cloudflare/next-on-pages";

Reviewer 2: Notes:
- We can make that available - maybe with a different name i.e. `getCloudflareContext` ?
- I think `process.env.MY_KV_NAME` would work

Author: If `process.env` would `just work` that would be awesome. cc @dario-piotrowicz

Reviewer: as discussed `process.env` would not just simply work we could look into that but I am not sure if that'd be that valuable `getCloudflareContext`'s ready to be used anyways 🙂

> // TODO: possible to bring this in from next-on-pages?
> import { getRequestContext } from "@cloudflare/next-on-pages";

Reviewer: ```suggestion
import { getCloudflareContext } from "@opennextjs/cloudflare";
```

export async function GET(request) {
	let responseText = "Hello World";

> 	const myKv = getRequestContext().env.MY_KV_NAMESPACE;

Reviewer: ```suggestion
	const myKv = (await getCloudflareContext()).env.MY_KV_NAMESPACE;
```

	await myKv.put("foo", "bar");
	const foo = await myKv.get("foo");

	return new Response(foo);
}
```

#### How to add bindings to your Worker

Add bindings to your Worker by [adding them to your `wrangler.toml` configuration file](/pages/functions/wrangler-configuration/).

## TypeScript type declarations for bindings

> To ensure that the `env` object from `getRequestContext().env` above has accurate TypeScript types, install [`@cloudflare/workers-types`](https://www.npmjs.com/package/@cloudflare/workers-types) and create a [TypeScript declaration file](https://www.typescriptlang.org/docs/handbook/2/type-declarations.html).

Reviewer: ```suggestion
To ensure that the `env` object from `(await getCloudflareContext()).env` above has accurate TypeScript types, install [`@cloudflare/workers-types`](https://www.npmjs.com/package/@cloudflare/workers-types) and create a [TypeScript declaration file](https://www.typescriptlang.org/docs/handbook/2/type-declarations.html).
```

Install Workers Types:

```sh
npm install --save-dev @cloudflare/workers-types
```

Add Workers Types to your `tsconfig.json` file, replacing the date below with your project's [compatibility date](https://developers.cloudflare.com/workers/configuration/compatibility-dates/):

```diff title="tsconfig.json"
    "types": [

> +        "@cloudflare/workers-types/2024-07-29"

Reviewer 2: Do we want the new dynamic types here?

Author: We should discuss — I'm inclined to say yes since this is pre 1.0 and the person using it is already using something new and bleeding edge?

    ]
```

> Create an `env.d.ts` file in the root directory of your Next.js app, and explicitly declare the type of each binding:

Author: Yeah, this is lifted from old docs, we should revise

```ts title="env.d.ts"
interface CloudflareEnv {
	MY_KV_1: KVNamespace;
	MY_KV_2: KVNamespace;
	MY_R2: R2Bucket;
	MY_DO: DurableObjectNamespace;
}
```

## Other Cloudflare APIs (`cf`, `ctx`)

> You can access context about the incoming request from the [`cf` object](https://developers.cloudflare.com/workers/runtime-apis/request/#incomingrequestcfproperties), as well as [lifecycle methods from the `ctx` object](https://developers.cloudflare.com/workers/runtime-apis/handlers/fetch/#lifecycle-methods) from the return value of [`getRequestContext()`](https://github.com/cloudflare/next-on-pages/blob/main/packages/next-on-pages/src/api/getRequestContext.ts):

Reviewer: ```suggestion
You can access context about the incoming request from the [`cf` object](https://developers.cloudflare.com/workers/runtime-apis/request/#the-cf-property-requestinitcfproperties), as well as lifecycle methods from the [`ctx` object](https://developers.cloudflare.com/workers/runtime-apis/context) from the return value of [`getCloudflareContext()`](https://github.com/opennextjs/opennextjs-cloudflare/blob/main/packages/cloudflare/src/api/get-cloudflare-context.ts):

```js

> // TODO: possible to bring this in from next-on-pages?
> import { getRequestContext } from "@cloudflare/next-on-pages";

Reviewer: ```suggestion
import { getCloudflareContext } from "@opennextjs/cloudflare";
```

> export const runtime = "edge";

Reviewer: ```suggestion
```

this is not necessary

export async function GET(request) {

> 	const { env, cf, ctx } = getRequestContext();

Reviewer: ```suggestion
	const { env, cf, ctx } = await getCloudflareContext();
```
